src/audio/drc/drc.c

```diff
@@ -140,353 +141,248 @@ static int drc_setup(struct drc_comp_data *cd, uint16_t channels, uint32_t rate)
+	comp_dbg(dev, "drc_process()");
 
-	cd->drc_func(dev, &source->stream, &sink->stream, frames);
+	/* Check for changed configuration */
+	if (comp_is_new_data_blob_available(cd->model_handler)) {
+		cd->config = comp_get_data_blob(cd->model_handler, NULL, NULL);
+		ret = drc_setup(cd, audio_stream_get_channels(sink),
+				audio_stream_get_rate(sink));
+		if (ret < 0) {
+			comp_err(dev, "drc_copy(), failed DRC setup");
+			return ret;
+		}
+	}
 
-	buffer_stream_writeback(sink, sink_bytes);
+	cd->drc_func(mod, source, sink, frames);
 
 	/* calc new free and available */
-	comp_update_buffer_consume(source, source_bytes);
-	comp_update_buffer_produce(sink, sink_bytes);
+	module_update_buffer_position(&input_buffers[0], &output_buffers[0], frames);
+	return 0;
 }
 
-/* copy and process stream data from source to sink buffers */
-static int drc_copy(struct comp_dev *dev)
+#if CONFIG_IPC_MAJOR_4
+static int drc_params(struct processing_module *mod)
 {
-	struct comp_copy_limits cl;
-	struct drc_comp_data *cd = comp_get_drvdata(dev);
-	struct comp_buffer *sourceb, *sinkb;
-	struct comp_buffer __sparse_cache *source_c, *sink_c;
-	int ret = 0;
+	struct sof_ipc_stream_params *params = mod->stream_params;
+	struct sof_ipc_stream_params comp_params;
+	struct comp_dev *dev = mod->dev;
+	struct comp_buffer *sinkb;
+	struct comp_buffer __sparse_cache *sink_c;
+	enum sof_ipc_frame valid_fmt, frame_fmt;
+	int i, ret;
 
-	comp_dbg(dev, "drc_copy()");
+	comp_dbg(dev, "drc_params()");
 
-	sourceb = list_first_item(&dev->bsource_list, struct comp_buffer,
-				  sink_list);
-	sinkb = list_first_item(&dev->bsink_list, struct comp_buffer,
-				source_list);
+	comp_params = *params;
+	comp_params.channels = mod->priv.cfg.base_cfg.audio_fmt.channels_count;
+	comp_params.rate = mod->priv.cfg.base_cfg.audio_fmt.sampling_frequency;
+	comp_params.buffer_fmt = mod->priv.cfg.base_cfg.audio_fmt.interleaving_style;
 
-	source_c = buffer_acquire(sourceb);
-	sink_c = buffer_acquire(sinkb);
+	audio_stream_fmt_conversion(mod->priv.cfg.base_cfg.audio_fmt.depth,
+				    mod->priv.cfg.base_cfg.audio_fmt.valid_bit_depth,
+				    &frame_fmt, &valid_fmt,
+				    mod->priv.cfg.base_cfg.audio_fmt.s_type);
 
-	/* Check for changed configuration */
-	if (comp_is_new_data_blob_available(cd->model_handler)) {
-		cd->config = comp_get_data_blob(cd->model_handler, NULL, NULL);
-		ret = drc_setup(cd, audio_stream_get_channels(&source_c->stream),
-				audio_stream_get_rate(&source_c->stream));
-		if (ret < 0) {
-			comp_err(dev, "drc_copy(), failed DRC setup");
-			goto out;
-		}
-	}
-
-	/* Get source, sink, number of frames etc. to process. */
-	comp_get_copy_limits(source_c, sink_c, &cl);
+	comp_params.frame_fmt = valid_fmt;
 
-	/* Run DRC function */
-	drc_process(dev, source_c, sink_c, cl.frames, cl.source_bytes,
-		    cl.sink_bytes);
+	for (i = 0; i < SOF_IPC_MAX_CHANNELS; i++)
+		comp_params.chmap[i] = (mod->priv.cfg.base_cfg.audio_fmt.ch_map >> i * 4) & 0xf;
```

Comment:
Can we get a helper for this? e.g. `get_chan_from_map(map, chan)`?

Reply:
I think I should convert this to use ipc4_update_buffer_format() from ipc/ipc4/helper.c, I've missed that such exists. It updates the channels map and everything else here.

Reply:
I've now updated it similarly as volume.c. The changed version updates both source and sink buffer parameters. I'm not sure about it so I'll check if that's the policy generally for simple 1 source 1 sink components. The patch to use sink parameters for DRC setup instead of source would be then unnecessary.